vespa-athenz/src/main/java/com/yahoo/vespa/athenz/client/zts/DefaultZtsClient.java

```diff
@@ -104,12 +120,64 @@ public InstanceIdentity refreshInstance(AthenzService providerIdentity,
         return withClient(client -> {
             try (CloseableHttpResponse response = client.execute(request)) {
                 return getInstanceIdentity(response);
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
             }
         });
     }
 
+    @Override
+    public ZToken getRoleToken(AthenzDomain domain) {
+        return getRoleToken(domain, null);
+    }
+
+    @Override
+    public ZToken getRoleToken(AthenzDomain domain, String roleName) {
+        URI uri = ztsUrl.resolve(String.format("domain/%s/token", domain.getName()));
+        RequestBuilder requestBuilder = RequestBuilder.get(uri)
+                .addHeader("Content-Type", "application/json");
+        if (roleName != null) {
+            requestBuilder.addParameter("role", roleName);
+        }
+        HttpUriRequest request = requestBuilder.build();
+        return withClient(client -> {
+            try (CloseableHttpResponse response = client.execute(request)) {
+                RoleTokenResponseEntity roleTokenResponseEntity = readEntity(response, RoleTokenResponseEntity.class);
+                return roleTokenResponseEntity.token;
+            }
+        });
+    }
+
+    @Override
+    public X509Certificate getRoleCertificate(AthenzDomain domain,
+                                              String roleName,
+                                              Duration expiry,
+                                              KeyPair keyPair,
+                                              String cloud) {
+        X500Principal principal = new X500Principal(String.format("cn=%s:role.%s", domain.getName(), roleName));
```

Comment:
I see roleName can be null with getRoleToken, but not with getRoleCertificate?

Reply:
True. A role certificate can only contain a single role, which must be specified. This is a limitation with certificates compared to role tokens.